src/fges.jl

```diff
@@ -0,0 +1,536 @@
+
+@memoize LRU(maxsize=1_000_000) function score(dataParsed::ParseData{Matrix{A}}, nodeParents, node) where A
+
+    # Unpack some variables from the dataParsed structure
+    n = A(dataParsed.numObservations) #convert datatype
+    data = dataParsed.normAugData
+    p = dataParsed.penalty
+
+    # The last column of dataParsed.normAugData is all ones which is required for a linear regression with an intercept. If there are no node parents, model the child node with just the intercept, else use the parents and the intercept
+    if isempty(nodeParents)
+        parentsAndIncept = [dataParsed.numFeatures+1]
+    else
+        parentsAndIncept = [nodeParents; dataParsed.numFeatures+1]
+    end
+
+    # To calculate the score we need a mean-squared error which we can get by regessing the the child node onto the parents
+    # Create variables for a linear regression y = X*b
+
+    # Use views to avoid creating copies of the data in memory
+        # X is the design matrix, augmented with a column of ones at the end
+        # X is also been standardized so mean(columns)=0 and std(column)=1
+        # y is data from the child node being tested
+    @views begin
+        y = data[:,node]
+        X = data[:,parentsAndIncept]
+    end
+
+    # Perform a linear regression
+    b = X \ y
+
+    # Get the estimation
+    ŷ = X*b
+
+    # Next we want to calculate the log likelihood that these are the parents of our node
+    # score = log(P(data|Model)) ≈ -BIC/2
+    # because we're only comparing log likelihoods we'll ignore the 1/2 factor
+    # when P(⋅) is Guassian, log(P(data|Model)) takes the form:
+    # -k⋅log(n) - n⋅log(mse)
+    # k is the number of free parameters and mse is mean squared error
+    k = length(parentsAndIncept) #includes the intercept
+    mse = sum(x->x^2, y-ŷ) / n
+
+    # return the final score we want to maximize (which is technically -2BIC)
+    return -p*k*log(n) - n*log(mse)
+end
+
+
+
+####################################################################
+# Using Meek's Rules to Update PDAG
+####################################################################
+
+# Revert a graph to undirected edges and unshielded colliders (i.e. parents not adjacent)
+function graphVStructure!(g)
+    
+    # loop through all vertices
+    for x in vertices(g)
+        # get the parents of current vertex
+        parentsX = parents(g,x)
+        # if all the parents are adjacent, undirect the edges
+        # (if there is only 1 parent, then it is still a clique)
+        if isclique(g, parentsX)
+            for p in parentsX
+                add_edge!(g, x, p)
+                add_edge!(g, p, x)
+            end
+        end
+    end
+end
+
+
+#Thoughts on improvement
+    #Would it be cleaner to find all unique triples where one edge is undirected?
+    #Can this update be more local? Most edges will remain unchanged.
+
+# This is set up so we only loop through all the vertices and edges once
```

Comment:
I don't think this works (at least if the edges are iterated over in arbitrary order):
E.g. if you have a - b - c - d <- e, then iterating over them in order a - b, b - c, etc will lead to a - b - c <- d <- e, I think. But it should be a <- b <- c <- d <- e.

Reply:
Agreed